src/base/memory_monitor_sr.hpp

```diff
@@ -0,0 +1,273 @@
+
+      static inline int MMM_MAKE_STAT_IDX_MAP (int comp_idx, int subcomp_idx)
+      {
+	return (comp_idx << 16 | subcomp_idx);
+      }
+
+      virtual ~mmm_module ()
+      {
+	if (m_module_name)
+	  {
+	    delete[] m_module_name;
+	  }
+      }
+    private:
+      /* private variable */
+      char *m_module_name;
+      MMM_MEM_STAT m_stat;
+      std::vector<mmm_component *> m_component;
+      std::vector<int> m_comp_idx_map;
+  };
+
+
+  /* you have to add your modules in Memory Monitoring Manager(MMM) class */
+  class memory_monitoring_manager
+  {
+    public:
+      memory_monitoring_manager (const char *name)
+      {
+	this->m_server_name = new char[strlen (name) + 1];
+	strcpy (this->m_server_name, name);
+	m_aggregater = new mmm_aggregater (this);
+      }
+
+
+      /* function */
+      int add_stat (THREAD_ENTRY *thread_p, MMM_STAT_ID stat_id, uint64_t size);
+      int sub_stat (THREAD_ENTRY *thread_p, MMM_STAT_ID stat_id, uint64_t size);
+      int resize_stat (THREAD_ENTRY *thread_p, MMM_STAT_ID stat_id, uint64_t old_size, uint64_t new_size);
+      int move_stat (THREAD_ENTRY *thread_p, MMM_STAT_ID src, MMM_STAT_ID dest, uint64_t size);
+      int aggregate_module_info (MEMMON_MODULE_INFO *info, int module_index);
+      int aggregate_tran_info (MEMMON_TRAN_INFO *info, int tran_count);
+
+      static int comp (const void *a, const void *b)
+      {
+	uint64_t x1 = ((std::pair<int, uint64_t> *)a)->second;
+	uint64_t x2 = ((std::pair<int, uint64_t> *)b)->second;
+
+	if (x1 < x2)
+	  {
+	    return 1;
+	  }
+	else if (x1 > x2)
+	  {
+	    return -1;
+	  }
+
+	return 0;
+      }
+      static inline int MMM_COMP_IDX_FROM_COMP_IDX_MAP (int comp_idx_map_val)
+      {
+	return (comp_idx_map_val >> 16);
+      }
+
+      static inline int MMM_SUBCOMP_IDX_FROM_COMP_IDX_MAP (int comp_idx_map_val)
+      {
+	return (comp_idx_map_val & MMM_PARSE_MASK);
+      }
+
+      ~memory_monitoring_manager ()
+      {
+	delete[] m_server_name;
+	for (int i = 0; i < MMM_MODULE_LAST; i++)
+	  {
+	    delete m_module[i];
+	  }
+      }
+    private:
+      class mmm_aggregater
+      {
+	public:
+	  mmm_aggregater (memory_monitoring_manager *mmm)
+	  {
+	    m_memmon_mgr = mmm;
+	  }
+
+	  /* function */
+	  int get_server_info (MEMMON_SERVER_INFO *info);
+	  int get_module_info (MEMMON_MODULE_INFO *info, int module_index);
+	  int get_transaction_info (MEMMON_TRAN_INFO *info, int tran_count);
+
+	  ~mmm_aggregater () {}
+	private:
+	  /* private variable */
+	  /* backlink of memory_monitoring_manager class */
+	  memory_monitoring_manager *m_memmon_mgr;
+      };
+
+    private:
+      /* private variable */
+      char *m_server_name;
+      std::atomic<uint64_t> m_total_mem_usage;
+      mmm_aggregater *m_aggregater;
+
+      /* Your module class should add in this array with print function */
+      mmm_module *m_module[MMM_MODULE_LAST] =
+      {
+	new mmm_module ()				  /* dummy for aligning module index (1 ~) */
+      };
```

Comment:
Move it to constructor. And how about set NULL if the dummy index is not used in your implementation?

Reply:
In my opinion, if we implement the definition in the constructor, we would need to declare each module individually like this:
m_module[0] = nullptr; (or m_module[0] = new mmm_module();)
m_module[1] = new mmm_heap_module("heap", heap_comp_info);
...
I don't think this approach is better than the current declaration and definition. If you have any good suggestions, please let me know.